ddtrace/contrib/flask/monkey.py

```diff
@@ -0,0 +1,487 @@
+            return traced_start_response
+        start_response = _wrap_start_response(start_response)
+
+        # DEV: We set response status code in `_wrap_start_response`
+        s.set_tag(http.URL, request.url)
+        s.set_tag(http.METHOD, request.method)
+
+        # TODO: Add request header tracing
+        # for k, v in request.headers:
+        #     s.set_tag('http.request.headers.{}'.format(k), v)
+        return wrapped(environ, start_response)
+
+
+def traced_blueprint_register(wrapped, instance, args, kwargs):
+    """
+    Wrapper for flask.blueprints.Blueprint.register
+
+    This wrapper just ensures the blueprint has a pin, either set manually on
+    itself from the user or inherited from the application
+    """
+    def _wrap(app, *args, **kwargs):
+        pin = Pin.get_from(instance)
+        if not pin:
+            pin = Pin.get_from(app)
+            if pin:
+                pin.clone().onto(instance)
+        return wrapped(app, *args, **kwargs)
+    return _wrap(*args, **kwargs)
+
+
+def traced_blueprint_add_url_rule(wrapped, instance, args, kwargs):
+    pin = get_inherited_pin(wrapped, instance)
+    if not pin:
+        return wrapped(*args, **kwargs)
+
+    def _wrap(rule, endpoint=None, view_func=None, **kwargs):
+        if view_func:
+            pin.clone().onto(view_func)
+        return wrapped(rule, endpoint=endpoint, view_func=view_func, **kwargs)
+
+    return _wrap(*args, **kwargs)
+
+
+def traced_add_url_rule(wrapped, instance, args, kwargs):
+    """Wrapper for flask.app.Flask.add_url_rule to wrap all views attached to this app"""
+    def _wrap(rule, endpoint=None, view_func=None, **kwargs):
+        if view_func:
+            # TODO: `if hasattr(view_func, 'view_class')` then this was generated from a `flask.views.View`
+            #   should we do something special with these views? Change the name/resource? Add tags?
+            view_func = wrap_function(instance, view_func, name=endpoint, resource=rule)
+
+        return wrapped(rule, endpoint=endpoint, view_func=view_func, **kwargs)
+
+    return _wrap(*args, **kwargs)
+
+
+def traced_endpoint(wrapped, instance, args, kwargs):
+    """Wrapper for flask.app.Flask.endpoint to ensure all endpoints are wrapped"""
+    def _wrap(endpoint):
+        def _wrapper(func):
+            name = func_name(func)
+            return wrapped(endpoint)(wrap_function(instance, func, name=name, resource=endpoint))
+        return _wrapper
+
+    return _wrap(*args, **kwargs)
+
+
+def traced_flask_hook(wrapped, instance, args, kwargs):
+    """Wrapper for hook functions (before_request, after_request, etc) are properly traced"""
+    def _wrap(func):
+        return wrapped(wrap_function(instance, func))
+
+    return _wrap(*args, **kwargs)
+
+
+def traced_render_template(wrapped, instance, args, kwargs):
+    """Wrapper for flask.templating.render_template"""
+    pin = get_inherited_pin(wrapped, instance, get_current_app())
```

Comment:
I see this in several places where we're checking `wrapped` for a pin instance. To my knowledge, we haven't really used functions to carry a pin.

I have nothing against it, in fact I think it makes a ton of sense for stand-alone functions.

We should consider making the `get_inherited_pin` function a standard helper function to be used across integrations. It would be good to get some consistency as to how to how to propagate the pin.